connect/api/src/main/java/org/apache/kafka/connect/header/ConnectHeaders.java

```diff
@@ -0,0 +1,534 @@
+                    }
+                }
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public Headers apply(HeaderTransform transform) {
+        if (!headers.isEmpty()) {
+            ListIterator<Header> iter = headers.listIterator();
+            while (iter.hasNext()) {
+                Header orig = iter.next();
+                Header updated = transform.apply(orig);
+                if (updated != null) {
+                    iter.set(updated);
+                } else {
+                    iter.remove();
+                }
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public int hashCode() {
+        return isEmpty() ? EMPTY_HASH : Objects.hash(headers);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) {
+            return true;
+        }
+        if (obj instanceof ConnectHeaders) {
+            ConnectHeaders that = (ConnectHeaders) obj;
+            if (this.isEmpty() && that.isEmpty()) {
+                return true;
+            }
+            return Objects.equals(this.headers, that.headers);
+        }
+        if (obj instanceof Headers) {
+            Headers that = (Headers) obj;
+            Iterator<Header> thisIter = this.iterator();
+            Iterator<Header> thatIter = that.iterator();
+            while (thisIter.hasNext() && thatIter.hasNext()) {
+                if (!Objects.equals(thisIter.next(), thatIter.next()))
+                    return false;
+            }
+            return !thisIter.hasNext() && !thatIter.hasNext();
+        }
+        return false;
+    }
+
+    @Override
+    public String toString() {
+        return "ConnectHeader(headers=" + (headers != null ? headers : "") + ")";
+    }
+
+    @Override
+    public ConnectHeaders duplicate() {
+        return new ConnectHeaders(this);
+    }
+
+    /**
+     * Check that the key is not null
+     *
+     * @param key the key; may not be null
+     * @throws IllegalArgumentException if the supplied key is null
+     */
+    private void checkKey(String key) {
+        if (key == null) {
+            throw new IllegalArgumentException("key cannot be null.");
+        }
+    }
+
+    /**
+     * Check the {@link Schema#type() schema's type} matches the specified type.
+     *
+     * @param schema the schema; never null
+     * @param type   the expected type
+     * @throws DataException if the schema's type does not match the expected type
+     */
+    private void checkSchemaType(Schema schema, Type type) {
+        if (schema.type() != type) {
+            throw new DataException("The schema for a list must be of type " + type);
+        }
+    }
+
+    /**
+     * Check that the value and its schema are compatible.
+     *
+     * @param schemaAndValue the schema and value pair
+     * @throws DataException if the schema is not compatible with the value
+     */
+    void checkSchemaMatches(SchemaAndValue schemaAndValue) {
```

Comment:
would SchemaAndValue be a better place to have this method?

Reply:
For some time it was not obvious to me that `SchemaAndValue` is not part of the public API, so my concern about that would be that it'd have to become a public method on `SchemaAndValue`, which is currently just a simple container. Here, we can keep it out of the API altogether.

Reply:
Got it.  May be we could have something like SchemaAndValueUtils to include such utils for it. Just  didn't feel great seeing this method in the ConnectHeaders class.

Reply:
`SchemaAndValue` is definitely public, it's used in some public APIs. Agree that there might be some handy utilities that could be colocated. I'd wait to put them there (and deal with any potential KIP fallout) until we know there are multiple use cases. atm, we don't have multiple use cases.

However, @rhauch, one question would be whether `ConnectSchema.validateValue` could work here instead of these methods?